### ISSUE TITLE
Windows Support

### ISSUE THREAD

**Issue by jennamagius** [oxy-secure/oxy#4]
Oxy should provide a bare minimum of Windows support for the things that are easy. This includes:

- BasicCommand

- File Transfer

- Port Forwards

Features that do not compile on Windows should be blocked out with conditional compilation guards, ensuring that these basic features continue to be usable on Windows.

Aspirationally, it would be great if an Oxy server running on windows could provide first-class interaction with a Windows Console object. This may be some-time off, however, as the Windows Console API is difficult to interact with and does not appear to provide an event-based mechanism for detecting and gathering Console output buffer changes as they occur.

**Comment by BlackHoleFox** [comment]
Has there been any progress towards client functionality for building on Windows?

**Comment by jennamagius** [comment]
Running a client on windows (outside of the linux subsystem)? Basically never, probably. Running a server on windows? Pretty straightforward, I just need to spend a day putting #[cfg()] guards around things Windows can't do. I'll probably go through and do a pass to get it building on windows in a week or two.

**Comment by jennamagius** [comment]
As of #42 and #65, it built and ran on Windows to the "easy" standard in the first post. Merely, since then I've added a whole bunch of stuff that needs new guards.

**Comment by BlackHoleFox** [comment]
So there will probably never be a client for Windows for Oxy, like there is Putty and OpenSSH for Windows?

**Comment by jennamagius** [comment]
Bulding one basically requires building a full terminal emulator because windows native terminal emulators are not capable of processing raw linux PTY output and displaying it usefully. Maybe by integrating with RXVT or something. It's not something I myself intend to work on in the foreseeable future.

I do intend to improve Windows server support.

**Comment by BlackHoleFox** [comment]
WSL it is then. I do look forward to being able to throw this on Windows Server though.

**Comment by BlackHoleFox** [comment]
@jennamagius Would it be possible to implement the switches so it could build on Windows properly, if it isn't too much that is? I would like to start helping with the projects development, and my only Linux machine is a POS laptop at the moment.

**Comment by jennamagius** [comment]
I'll do it Saturday, but I'll just tell you: developing on this project on Windows is going to be a nightmare for you. I'll get it to build fairly quickly, but you're going to be faced with a huge number of "dead code" lints that only fire because it's dead code on Windows, not other platforms. Plus, like 80% of the lines of code you'll be seeing won't be relevant to you.

I really appreciate your desire to contribute! Really, the best contribution to be made from Windows right now would be getting it to compile on Windows. >.> The work pretty much involves putting #[cfg(unix)] around huge swaths of code until all the parts that don't work on windows are hidden from the compiler. Maybe a few unimplemented!()s to get around expected return types.

**Comment by BlackHoleFox** [comment]
I can probably poke at that tomorrow. Also gonna look at getting dual booting going on my desktop. Any recommendations for Linux distros? Only ever really used headless versions so not sure whats good with GUIs.

**Comment by BlackHoleFox** [comment]
Ubuntu it is then, or some fork of it. I have a fair bit of experience with it as around 70% of the servers I manage run something either Linux or BSD, so I should be good with no machine bricking. Just new to the graphical Linux distros.

**Comment by jennamagius** [comment]
Turns out I lied: I'm definitely going to spend next weekend ripping out and replacing the whole crypto stack with Noise.

Probably the weekend after that `>.>`

**Comment by BlackHoleFox** [comment]
Noise is random generation type stuff right? Isn't Ring in the project for crypto stuff as well?

**Comment by jennamagius** [comment]
Nah, it's actually a key exchange/handshake specification. Ring is potentially still the underlying crypto provider, but there will be another library (Snow) that sits between Oxy and Ring and manages the key exchange

**Comment by BlackHoleFox** [comment]
Oh cool. I've looked at Ring before but not those. I'll have to look over the docs some and see what they can do.

**Comment by jennamagius** [comment]
https://noiseprotocol.org/noise.html

We're going to wind up using an IKpsk1 type handshake. (specifically, Noise_IKpsk1_25519_AESGCM_SHA512). The spec is a lot to read, but I think it makes the crypto significantly better, ultimately.

**Comment by jennamagius** [comment]
As of https://github.com/oxy-secure/oxy/commit/478e51a76560611b93f8f20aafb1375edbb70fbc , it should _compile_ on Windows. There'll be a huge amount of spurious lints, and the only useful modes will be "serve-one" and "reverse-server". Even then, after connecting to a windows server there's no directly interactive terminal. Merely, you can hit F10 and run "sh whoami" type BasicCommands from the metacommand prompt.

But hey, it compiles.

**Comment by jennamagius** [comment]
Also portforwards and file transfer should work just fine against a Windows server.
